pkg/kubectl/rolling_updater_test.go

```diff
@@ -131,396 +78,479 @@ func newRc(replicas int, desired int) *api.ReplicationController {
+				up{3},
+				down{oldReady: 10, newReady: 3, to: 7},
+				up{6},
+				down{oldReady: 7, newReady: 6, to: 4},
+				up{9},
+				down{oldReady: 4, newReady: 9, to: 1},
+				up{10},
+				down{oldReady: 1, newReady: 10, to: 0},
 			},
-			output: `Creating foo-v2
-Scaling up foo-v2 from 0 to 7, scaling down foo-v1 from 2 to 0 (scale up first by 1 each interval)
-Scaling foo-v2 up to 1
+			output: `Created foo-v2
+Scaling up foo-v2 from 0 to 10, scaling down foo-v1 from 10 to 0 (max unavailable 0, max surge 3)
+Scaling foo-v2 up to 3
+Scaling foo-v1 down to 7
+Scaling foo-v2 up to 6
+Scaling foo-v1 down to 4
+Scaling foo-v2 up to 9
 Scaling foo-v1 down to 1
-Scaling foo-v2 up to 2
+Scaling foo-v2 up to 10
 Scaling foo-v1 down to 0
-Scaling foo-v2 up to 7
-Update succeeded. Deleting foo-v1
 `,
 		}, {
-			oldRc:    oldRc(7),
-			newRc:    newRc(2, 2),
-			accepted: true,
-			percent:  NilPercent,
-			responses: []fakeResponse{
-				// no existing newRc
-				{nil, fmt.Errorf("not found")},
-				// scaling iteration
-				{newRc(1, 2), nil},
-				{oldRc(6), nil},
-				// scaling iteration
-				{newRc(2, 2), nil},
-				{oldRc(0), nil},
-				// cleanup annotations
-				{newRc(2, 2), nil},
-				{newRc(2, 2), nil},
-				{newRc(2, 2), nil},
+			name:        "10->10 0/30 delayed readiness",
+			oldRc:       oldRc(10, 10),
+			newRc:       newRc(0, 10),
+			newRcExists: false,
+			maxUnavail:  0,
+			maxSurge:    30,
+			expected: []interface{}{
+				up{3},
+				down{oldReady: 10, newReady: 0, noop: true},
+				down{oldReady: 10, newReady: 1, to: 9},
+				up{4},
+				down{oldReady: 9, newReady: 3, to: 7},
+				up{6},
+				down{oldReady: 7, newReady: 6, to: 4},
+				up{9},
+				down{oldReady: 4, newReady: 9, to: 1},
+				up{10},
+				down{oldReady: 1, newReady: 9, noop: true},
+				down{oldReady: 1, newReady: 10, to: 0},
 			},
-			output: `Creating foo-v2
-Scaling up foo-v2 from 0 to 2, scaling down foo-v1 from 7 to 0 (scale up first by 1 each interval)
-Scaling foo-v2 up to 1
-Scaling foo-v1 down to 6
-Scaling foo-v2 up to 2
+			output: `Created foo-v2
+Scaling up foo-v2 from 0 to 10, scaling down foo-v1 from 10 to 0 (max unavailable 0, max surge 3)
+Scaling foo-v2 up to 3
+Scaling foo-v1 down to 9
+Scaling foo-v2 up to 4
+Scaling foo-v1 down to 7
+Scaling foo-v2 up to 6
+Scaling foo-v1 down to 4
+Scaling foo-v2 up to 9
+Scaling foo-v1 down to 1
+Scaling foo-v2 up to 10
 Scaling foo-v1 down to 0
-Update succeeded. Deleting foo-v1
 `,
 		}, {
-			oldRc:    oldRc(7),
-			newRc:    newRc(2, 2),
-			accepted: false,
-			percent:  NilPercent,
-			responses: []fakeResponse{
-				// no existing newRc
-				{nil, fmt.Errorf("not found")},
-				// scaling iteration (only up occurs since the update is rejected)
-				{newRc(1, 2), nil},
+			name:        "10->10 10/20 fast readiness",
+			oldRc:       oldRc(10, 10),
+			newRc:       newRc(0, 10),
+			newRcExists: false,
+			maxUnavail:  10,
+			maxSurge:    20,
+			expected: []interface{}{
+				up{3},
```

Comment:
I implemented what I think is a fix for this. When setting up to start the loop, if old+new replicas is > desired, we know we can skip the first scale up and get it moving again.

When scaling down, if we compute there's no way to scale down without minimum violation but the availability counts match their corresponding old/new replica counts, and we're not already at the desired, we know there are no more pods forthcoming (i.e. we missed a scale-up). So, just skip scale down and return the max scale up value to get things moving upwards again.